x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineTests.java

```diff
@@ -571,24 +577,33 @@ public void testProcessOnceOnPrimary() throws Exception {
             }
         }
         Randomness.shuffle(operations);
+        final long oldTerm = randomLongBetween(1, Integer.MAX_VALUE);
+        primaryTerm.set(oldTerm);
         try (Store store = createStore(shardId, indexSettings, newDirectory())) {
             final EngineConfig engineConfig = engineConfig(shardId, indexSettings, threadPool, store, logger, xContentRegistry());
             try (FollowingEngine followingEngine = createEngine(store, engineConfig)) {
                 followingEngine.advanceMaxSeqNoOfUpdatesOrDeletes(operations.size() - 1L);
-                final long oldTerm = randomLongBetween(1, Integer.MAX_VALUE);
+                final Map<Long,Long> operationWithTerms = new HashMap<>();
                 for (Engine.Operation op : operations) {
-                    Engine.Result result = applyOperation(followingEngine, op, oldTerm, randomFrom(Engine.Operation.Origin.values()));
+                    long term = randomLongBetween(1, oldTerm);
+                    Engine.Result result = applyOperation(followingEngine, op, term, randomFrom(Engine.Operation.Origin.values()));
                     assertThat(result.getResultType(), equalTo(Engine.Result.Type.SUCCESS));
+                    operationWithTerms.put(op.seqNo(), term);
+                    if (rarely()) {
+                        followingEngine.refresh("test");
+                    }
                 }
                 // Primary should reject duplicates
                 final long newTerm = randomLongBetween(oldTerm + 1, Long.MAX_VALUE);
                 for (Engine.Operation op : operations) {
                     Engine.Result result = applyOperation(followingEngine, op, newTerm, Engine.Operation.Origin.PRIMARY);
                     assertThat(result.getResultType(), equalTo(Engine.Result.Type.FAILURE));
                     assertThat(result.getFailure(), instanceOf(AlreadyProcessedFollowingEngineException.class));
+                    AlreadyProcessedFollowingEngineException failure = (AlreadyProcessedFollowingEngineException) result.getFailure();
+                    assertThat(failure.getExistingPrimaryTerm().getAsLong(), equalTo(operationWithTerms.get(op.seqNo())));
```

Comment:
this seems to mean we never deliver ops below the global checkpoint (and flush etc.) can we extend the test to that too?